subprojects/dependency-management/src/main/java/org/gradle/internal/component/model/DefaultComponentGraphResolveState.java

```diff
@@ -63,16 +65,17 @@ public ComponentArtifactMetadata resolveArtifact(IvyArtifactName artifact) {
 
         @Override
         public ArtifactSet resolveArtifacts(ArtifactSelector artifactSelector, ExcludeSpec exclusions, ImmutableAttributes overriddenAttributes) {
-            // We do not currently cache ResolvedVariants for external modules
-            return artifactSelector.resolveArtifacts(artifactMetadata, null, this::buildAllVariants, graphSelectedVariant.getVariants(), exclusions, overriddenAttributes);
+            Set<? extends VariantResolveMetadata> fallbackVariants = graphSelectedVariant.getVariants();
+            Optional<List<? extends VariantGraphResolveMetadata>> variantsForGraphTraversal = graphMetadata.getVariantsForGraphTraversal();
+            return artifactSelector.resolveArtifacts(artifactMetadata, () -> buildAllVariants(fallbackVariants, variantsForGraphTraversal), fallbackVariants, exclusions, overriddenAttributes);
         }
 
-        private Set<? extends VariantResolveMetadata> buildAllVariants() {
+        private static Set<? extends VariantResolveMetadata> buildAllVariants(Set<? extends VariantResolveMetadata> fallbackVariants, Optional<List<? extends VariantGraphResolveMetadata>> variantsForGraphTraversal) {
             final Set<? extends VariantResolveMetadata> allVariants;
-            if (graphMetadata.getVariantsForGraphTraversal().isPresent()) {
-                allVariants = graphMetadata.getVariantsForGraphTraversal().get().stream().map(ModuleConfigurationMetadata.class::cast).flatMap(variant -> variant.getVariants().stream()).collect(Collectors.toSet());
+            if (variantsForGraphTraversal.isPresent()) {
+                allVariants = variantsForGraphTraversal.get().stream().map(ModuleConfigurationMetadata.class::cast).flatMap(variant -> variant.getVariants().stream()).collect(Collectors.toSet());
             } else {
-                allVariants = graphSelectedVariant.getVariants();
+                allVariants = fallbackVariants;
             }
             return allVariants;
         }
```

Comment:
🤔 This is very similar to the `buildAllVariants` method in `DefaultLocalComponentGraphResolveState`.  You could extract a common method like:

```java
private static Set<? extends VariantResolveMetadata> buildAllVariants(Set<? extends VariantResolveMetadata> fallbackVariants, Optional<List<? extends VariantGraphResolveMetadata>> variantsForGraphTraversal, Function<VariantGraphResolveMetadata, ConfigurationMetadata> prepMethod) {
    final Set<? extends VariantResolveMetadata> allVariants;
    if (variantsForGraphTraversal.isPresent()) {
        allVariants = variantsForGraphTraversal.get().stream().
            map(VariantGraphResolveMetadata.class::cast).
            map(prepMethod).
            flatMap(variant -> variant.getVariants().stream()).
            collect(Collectors.toSet());
    } else {
        allVariants = fallbackVariants;
    }
    return allVariants;
}
```

And move it to `AbstractComponentGraphResolveState`.